Name resolution problem with non-top level functions

This is a major problem that is blocking the 230 Final :( 

see `tests/todo/MM.hs` (which imports `tests/todo/LL.hs`) and fails with:

```
 /Users/rjhala/research/liquid/liquidhaskell/tests/todo/LL.hs:8:9: Error: GHC Error
     Not in scope: variable `LL.go'
```

(the problem goes away if I just make `go` a top-level function.)

I'm guessing this has to do with @spinda's recent changes?

--- a/src/Language/Haskell/Liquid/Bare/Check.hs
+++ b/src/Language/Haskell/Liquid/Bare/Check.hs
@@ -5,7 +5,6 @@
 module Language.Haskell.Liquid.Bare.Check (
     checkGhcSpec
 
-  , checkAppTys
   , checkTerminationExpr
   , checkTy
   ) where

--- a/src/Language/Haskell/Liquid/Bare/Check.hs
+++ b/src/Language/Haskell/Liquid/Bare/Check.hs
@@ -6,7 +6,6 @@ module Language.Haskell.Liquid.Bare.Check (
     checkGhcSpec
 
   , checkAppTys
-  , checkDefAsserts
   , checkTerminationExpr
   , checkTy
   ) where
@@ -23,14 +22,11 @@ import Control.Arrow ((&&&))
 import Control.Monad.Writer
 import Data.Maybe
 import Text.PrettyPrint.HughesPJ
-import Text.Printf
 
 import qualified Data.List           as L
 import qualified Data.HashMap.Strict as M
-import qualified Data.HashSet        as S
 
 import Language.Fixpoint.Misc (applyNonNull, group, mapSnd, snd3, errorstar, safeHead)
-import Language.Fixpoint.Names (isPrefixOfSym, stripParensSym)
 import Language.Fixpoint.Sort (checkSorted, checkSortedReftFull, checkSortFull)
 import Language.Fixpoint.Types hiding (R)
 
@@ -344,27 +340,4 @@ checkMBody' emb sort γ body = case body of
                                  , ty_binds = tail $ ty_binds trep'
                                  , ty_args  = tail $ ty_args trep'             }
     trep' = toRTypeRep sort
-
-
-checkDefAsserts :: BareEnv -> [(Var, LocSymbol, BareType)] -> [(LocSymbol, BareType)] -> BareM ()
-checkDefAsserts env vbs xbs   = applyNonNull (return ()) grumble  undefSigs
-  where
-    undefSigs                 = [x | (x, _) <- assertSigs, not (x `S.member` definedSigs)]
-    assertSigs                = filter isTarget xbs
-    definedSigs               = S.fromList $ snd3 <$> vbs
-    grumble                   = mapM_ (warn . berrUnknownVar)
-    moduleName                = symbol $ modName env
-    isTarget                  = isPrefixOfSym moduleName . stripParensSym . val . fst
-
-warn x = tell [x]
-
--------------------------------------------------------------------------------------
--- | Tasteful Error Messages --------------------------------------------------------
--------------------------------------------------------------------------------------
-
-berrUnknownVar       = berrUnknown "Variable"
-
-berrUnknown :: (PPrint a) => String -> Located a -> String 
-berrUnknown thing x  = printf "[%s]\nSpecification for unknown %s : %s"  
-                         thing (showpp $ loc x) (showpp $ val x)
 

--- a/src/Language/Haskell/Liquid/Bare/GhcSpec.hs
+++ b/src/Language/Haskell/Liquid/Bare/GhcSpec.hs
@@ -105,7 +105,7 @@ makeGhcSpec' cfg cbs vars defVars exports specs
        makeRTEnv specs
        (tycons, datacons, dcSs, tyi, embs)     <- makeGhcSpecCHOP1 specs
        modify                                   $ \be -> be { tcEnv = tyi }
-       (cls, mts)                              <- second mconcat . unzip . mconcat <$> mapM (makeClasses cfg vars) specs
+       (cls, mts)                              <- second mconcat . unzip . mconcat <$> mapM (makeClasses name cfg vars) specs
        (measures, cms', ms', cs', xs')         <- makeGhcSpecCHOP2 cbs specs dcSs datacons cls embs
        (invs, ialias, sigs, asms)              <- makeGhcSpecCHOP3 cfg vars defVars specs name mts embs
        syms                                    <- makeSymbols (vars ++ map fst cs') xs' (sigs ++ asms ++ cs') ms' (invs ++ (snd <$> ialias))

--- a/src/Language/Haskell/Liquid/Bare/Spec.hs
+++ b/src/Language/Haskell/Liquid/Bare/Spec.hs
@@ -122,13 +122,13 @@ makeAssertSpec cmod cfg vs lvs (mod,spec)
   | cmod == mod
   = makeLocalSpec cfg cmod vs lvs (grepClassAsserts (Ms.rinstance spec)) (Ms.sigs spec ++ Ms.localSigs spec)
   | otherwise
-  = inModule mod $ makeSpec cfg vs $ Ms.sigs spec
+  = return []
 
 makeAssumeSpec cmod cfg vs lvs (mod,spec)
   | cmod == mod
   = makeLocalSpec cfg cmod vs lvs [] $ Ms.asmSigs spec
   | otherwise
-  = inModule mod $ makeSpec cfg vs $ Ms.asmSigs spec
+  = inModule mod $ makeSpec cfg vs (Ms.sigs spec ++ Ms.asmSigs spec)
 
 grepClassAsserts  = concatMap go 
    where

--- a/src/Language/Haskell/Liquid/Bare/Spec.hs
+++ b/src/Language/Haskell/Liquid/Bare/Spec.hs
@@ -123,13 +123,13 @@ makeAssertSpec cmod cfg vs lvs (mod,spec)
   | cmod == mod
   = makeLocalSpec cfg cmod vs lvs (grepClassAsserts (Ms.rinstance spec)) (Ms.sigs spec ++ Ms.localSigs spec)
   | otherwise
-  = return []
+  = inModule mod $ makeSpec True vs $ Ms.sigs spec
 
 makeAssumeSpec cmod cfg vs lvs (mod,spec)
   | cmod == mod
   = makeLocalSpec cfg cmod vs lvs [] $ Ms.asmSigs spec
   | otherwise
-  = inModule mod $ makeSpec True vs (Ms.sigs spec ++ Ms.asmSigs spec)
+  = inModule mod $ makeSpec True vs $ Ms.asmSigs spec
 
 grepClassAsserts  = concatMap go 
    where

--- a/src/Language/Haskell/Liquid/Bare/Spec.hs
+++ b/src/Language/Haskell/Liquid/Bare/Spec.hs
@@ -57,7 +57,7 @@ import Language.Haskell.Liquid.Bare.OfType
 import Language.Haskell.Liquid.Bare.Resolve
 import Language.Haskell.Liquid.Bare.SymSort
 
-makeClasses cfg vs (mod, spec) = inModule mod $ mapM mkClass $ Ms.classes spec
+makeClasses cmod cfg vs (mod, spec) = inModule mod $ mapM mkClass $ Ms.classes spec
   where
     --FIXME: cleanup this code
     unClass = snd . bkClass . fourth4 . bkUniv
@@ -69,7 +69,7 @@ makeClasses cfg vs (mod, spec) = inModule mod $ mapM mkClass $ Ms.classes spec
                  let αs  = map symbolRTyVar as
                  let as' = [rVar $ symbolTyVar a | a <- as ]
                  let ms' = [ (s, rFun "" (RApp c (flip RVar mempty <$> as) [] mempty) t) | (s, t) <- ms]
-                 vts <- makeSpec (noCheckUnknown cfg) vs ms'
+                 vts <- makeSpec (noCheckUnknown cfg || cmod /= mod) vs ms'
                  let sts = [(val s, unClass $ val t) | (s, _)    <- ms
                                                      | (_, _, t) <- vts]
                  let t   = rCls tc as'
@@ -129,7 +129,7 @@ makeAssumeSpec cmod cfg vs lvs (mod,spec)
   | cmod == mod
   = makeLocalSpec cfg cmod vs lvs [] $ Ms.asmSigs spec
   | otherwise
-  = inModule mod $ makeSpec (noCheckUnknown cfg) vs (Ms.sigs spec ++ Ms.asmSigs spec)
+  = inModule mod $ makeSpec True vs (Ms.sigs spec ++ Ms.asmSigs spec)
 
 grepClassAsserts  = concatMap go 
    where

--- a/src/Language/Haskell/Liquid/Bare/Spec.hs
+++ b/src/Language/Haskell/Liquid/Bare/Spec.hs
@@ -23,10 +23,12 @@ module Language.Haskell.Liquid.Bare.Spec (
   , makeSpecDictionaries
   ) where
 
+import MonadUtils (mapMaybeM)
 import TyCon
 import Var
 
 import Control.Applicative ((<$>))
+import Control.Monad.Error
 import Control.Monad.State
 import Data.Maybe
 import Data.Monoid
@@ -47,7 +49,6 @@ import Language.Haskell.Liquid.Types
 
 import qualified Language.Haskell.Liquid.Measure as Ms
 
-import Language.Haskell.Liquid.Bare.Check (checkDefAsserts)
 import Language.Haskell.Liquid.Bare.Env
 import Language.Haskell.Liquid.Bare.Existential
 import Language.Haskell.Liquid.Bare.Lookup
@@ -68,7 +69,7 @@ makeClasses cfg vs (mod, spec) = inModule mod $ mapM mkClass $ Ms.classes spec
                  let αs  = map symbolRTyVar as
                  let as' = [rVar $ symbolTyVar a | a <- as ]
                  let ms' = [ (s, rFun "" (RApp c (flip RVar mempty <$> as) [] mempty) t) | (s, t) <- ms]
-                 vts <- makeSpec cfg vs ms'
+                 vts <- makeSpec (noCheckUnknown cfg) vs ms'
                  let sts = [(val s, unClass $ val t) | (s, _)    <- ms
                                                      | (_, _, t) <- vts]
                  let t   = rCls tc as'
@@ -128,7 +129,7 @@ makeAssumeSpec cmod cfg vs lvs (mod,spec)
   | cmod == mod
   = makeLocalSpec cfg cmod vs lvs [] $ Ms.asmSigs spec
   | otherwise
-  = inModule mod $ makeSpec cfg vs (Ms.sigs spec ++ Ms.asmSigs spec)
+  = inModule mod $ makeSpec (noCheckUnknown cfg) vs (Ms.sigs spec ++ Ms.asmSigs spec)
 
 grepClassAsserts  = concatMap go 
    where
@@ -153,11 +154,9 @@ makeDefaultMethods defVs sigs
 makeLocalSpec :: Config -> ModName -> [Var] -> [Var] -> [(LocSymbol, BareType)] -> [(LocSymbol, BareType)]
                     -> BareM [(ModName, Var, Located SpecType)]
 makeLocalSpec cfg mod vs lvs cbs xbs
-  = do env   <- get
-       vbs1  <- fmap expand3 <$> varSymbols fchoose lvs (dupSnd <$> xbs1)
-       unless (noCheckUnknown cfg)   $ checkDefAsserts env vbs1 xbs1
+  = do vbs1  <- fmap expand3 <$> varSymbols fchoose lvs (dupSnd <$> xbs1)
        vts1  <- map (addFst3 mod) <$> mapM mkVarSpec vbs1
-       vts2  <- makeSpec cfg vs xbs2
+       vts2  <- makeSpec (noCheckUnknown cfg) vs xbs2
        return $ vts1 ++ vts2
   where
     xbs1 = xbs1' ++ cbs
@@ -168,18 +167,24 @@ makeLocalSpec cfg mod vs lvs cbs xbs
     fchoose ls          = maybe ls (:[]) $ L.find (`elem` vs) ls
     modElem n x         = (takeModuleNames $ val x) == (symbol n)
 
-makeSpec :: Config -> [Var] -> [(LocSymbol, BareType)]
-                -> BareM [(ModName, Var, Located SpecType)]
-makeSpec cfg vs xbs
-  = do vbs <- map (joinVar vs) <$> lookupIds xbs
-       env@(BE { modName = mod}) <- get
-       unless (noCheckUnknown cfg) $ checkDefAsserts env vbs xbs
+makeSpec :: Bool -> [Var] -> [(LocSymbol, BareType)]
+                 -> BareM [(ModName, Var, Located SpecType)]
+makeSpec ignoreUnknown vs xbs
+  = do vbs <- map (joinVar vs) <$> lookupIds ignoreUnknown xbs
+       (BE { modName = mod}) <- get
        map (addFst3 mod) <$> mapM mkVarSpec vbs
 
 
-lookupIds = mapM lookup
+lookupIds ignoreUnknown
+  = mapMaybeM lookup
   where
-    lookup (s, t) = (,s,t) <$> lookupGhcVar s
+    lookup (s, t)
+      = (Just . (,s,t) <$> lookupGhcVar s) `catchError` handleError
+    handleError (ErrGhc {})
+      | ignoreUnknown
+        = return Nothing
+    handleError err
+      = throwError err
 
 mkVarSpec :: (Var, LocSymbol, BareType) -> BareM (Var, Located SpecType)
 mkVarSpec (v, Loc l _, b) = tx <$> mkSpecType l b

--- a/tests/pos/LocalSpec0.hs
+++ b/tests/pos/LocalSpec0.hs
@@ -1,4 +1,4 @@
-module LL (foo) where
+module LocalSpec0 (foo) where
 
 
 {-@ foo :: x:Int -> {v:Int | v > x } @-}

--- a/tests/pos/LocalSpecImp.hs
+++ b/tests/pos/LocalSpecImp.hs
@@ -1,0 +1,7 @@
+module LocalSpecImp where
+
+import LocalSpec0
+
+{-@ bar :: {x:Int | x > 99} -> {v:Int | v > 100 } @-}
+bar :: Int -> Int
+bar x = foo x

--- a/tests/todo/MM.hs
+++ b/tests/todo/MM.hs
@@ -1,7 +1,0 @@
-module MM where
-
-import LL
-
-{-@ bar :: x:Int -> {v:Int | v > 100 } @-}
-bar :: Int -> Int
-bar x = foo x

